### ISSUE TITLE
Editors without publish permission can easily change images on live pages.

### ISSUE THREAD

**Issue by HsnVahedi** [wagtail/wagtail#6594]
### Issue Summary

Editors without publish permission can easily change images on live pages.

### Steps to Reproduce

1. Suppose we have a page defined like this:

```python
class SamplePage(Page):
    image = models.ForeignKey(
        'wagtailimages.Image',
        null=True,
        blank=True,
        on_delete=models.SET_NULL,
        related_name='+'
    )

    content_panels = Page.content_panels + [
        ImageChooserPanel('image')
    ]
```

2. As an administrator, create a new group with permission to add (without publish) pages in a particular directory.
3. Also, give this created group, the permission to add images.
4. Create a new user in that group.
5. Login as the created user and upload an image and create a SamplePage and submit it for moderation.
6. As an administrator, approve it and publish the new SamplePage.
7. Login as the created user again, Now you can easily change the image on the live SamplePage !!!! And no moderator can prevent you to do that !!!!! In other words, you can ruin the whole page when it is live.

This is a real bug I think. Because, as an administrator, I don't want people without publish permission, to be able to change the live pages whenever they want.

**Comment by tomdyson** [comment]
Thanks for the feedback, @HsnVahedi. What do you think the behaviour should be? Separate add and edit permissions for images?

**Comment by HsnVahedi** [comment]
Thanks a lot for your response, sir! @tomdyson. What do you mean by "Separate and edit permissions for images"?

Personally I believe storing foreign keys in a Page object is not a safe approach. Pages should be able to store values (copies) and maintain snapshots. I think it's possible, isn't it?

**Comment by HsnVahedi** [comment]
I'm wondering if a wrapper class could solve this problem. something like this:
```

class PageImage(Snapshotable):
    class Meta:
        model_class_name = "wagtailimages.Image"

class SamplePage(Page):
    image = models.OneToOneField(PageImage)
```

**Comment by jacobtoppm** [comment]
Hi @HsnVahedi, thanks for bringing this up. We've been discussing this in the core team meeting, and I wanted to let you know where we got to. Basically, snapshotting objects/preventing edits via permissions as you suggested on the other issue aren't always desired behaviours: often the power of Snippets and Images comes in being able to edit and update them centrally. However, it's definitely a valid concern that editors without permissions can make changes to published pages, so these kinds of checks are something we'd be interested in as a (non-default) option for developers. 

However, in order to adopt these checks, we'd need a way of efficiently finding "pages that reference this object", as object references can be included not only via straightforward `ForeignKeys`, but also within `StreamField` content or rich text. This is something that is currently on the [Wagtail Roadmap](https://github.com/wagtail/wagtail/wiki/Roadmap), aiming for Q3 2021. After we've got efficient object reference finding, implementing something like this would be far easier. There was some discussion of starting with a limited, `ForeignKey`-only version of this, but currently we don't have development time. If this or object finding is something you're interested in, we'd definitely encourage any contributions!

**Comment by HsnVahedi** [comment]
Thanks for letting me know about your discussion. @jacobtoppm 

I'll let you know if I could help.

**Comment by ababic** [comment]
I wonder if we might be overcomplicating this. Aren't there separate 'add' and 'edit' permissions for collections already? And if users only have the 'add' permission, that removed the ability to edit? If so, isn't this just a case of assigning group permissions in such a way that only users you trust receive the 'edit' permission?

**Comment by HsnVahedi** [comment]
@ababic
Thanks for contributing!

`Aren't there separate 'add' and 'edit' permissions for collections already?` There are seperate add and edit permissions but when you add something, you automatically gain permission to edit it.

`isn't this just a case of assigning group permissions in such a way that only users you trust receive the 'edit' permission?` Unfortunately no. Because in this scenario, permission on editing the image has to get dynamically computed according to not only the image but also the pages that reference it.

**Comment by ababic** [comment]
Generally, I think projects benifit from the ability to manage media centrally. For example: An uploaded image mistakenly includes a brand logo and needs replacing with a version that has the logo removed. Currently, that would be an easy thing to fix, but what would that process look like if page content was tied to older versions?

Likewise, a site has several pages linking to a 'terms and conditions' PDF that is revised every 6 months or so. This kind of update is usually time-sensitive, so waiting on all content editors to find and update all pages to link to the new version wouldn't be suitable.

Needs will always differ from project-to-project, and team size makes quite a big difference. For example, If there are only couple of CMS users taking care of everything, this is unlikely to be a problem. Even with a larger team, where one or two user might ONLY have media permissions, I think the likelihood of this being a problem is quite small if those users are trained and trusted. @HsnVahedi, in your use case, what is the real concern? Rouge users potentially updating images/documents with malicious intentent, or is it more about limiting damage of maliciously hijacked accounts? It would be good to understand more about the problem before we get too attached to solutions.

I will admit, I kind of like the idea of an interim model to support tying content revisions to specific versions of media, But that does bring with it its own set of complications. For example:

1. How would we make content editors aware that they are using an 'old' versions of an image/document  somewhere? How would we encourage them to keep content up-to-date (if at all)? 
2. How would image revision history be managed (in UI terms)? I've seen sites with 5,000+ images in the library. Keeping every version (original + all revisions) of every image, both the database and in file storage, could easily become burdensome.
3. How would we cater for cases where centrally managed images/documents would be preferred? What would the UI be like? Would that be specified for each use, or would it be set on the image/document itself? 

It's worth highlighting that large features/changes like this tend not to be implemented unless sponsored by a client, or if developed by a third-party and merged in later. Either way, the best route is almost always to create a third-party app first, where things can evolve organically, not tied to Wagtail's release process. @HsnVahedi, if you have avenues you'd like to explore, may I suggest trying this approach?

**Comment by HsnVahedi** [comment]
Thanks a lot for sharing your concerns. @ababic

You are completely right about the importance of being able to manage media centrally. It has to remain in wagtail.

But I beileve giving developers the ability to use a new way of managing media for some of their images/documents/... fields, makes wagtail really powerful.

> in your use case, what is the real concern?

I want to have a complete CMS in which a lot of editors (unknown people who moderators might not trust) could create content.
I don't want any of the editors to put some image on a live web page without moderators approval.

> 1. How would we make content editors aware that they are using an 'old' version of an image/document somewhere? How would we encourage them to keep content up-to-date (if at all)?

> 2. How would image revision history be managed (in UI terms)? I've seen sites with 5,000+ images in the library. Keeping every version (original + all revisions) of every image, both in the database and in file storage, could easily become a burden.

I think answering to these questions is really hard (even if I would have understood them correctly). As you've mentioned, keeping versions of images/documents/... is really hard and complex. I have discussed a much simpler approach [in this issue](https://github.com/wagtail/wagtail/issues/6600#issuecomment-742423966) which doesn't require keeping versions.

> How would we also support simpler use-cases where centrally managed images/documents is preferred? 

I do believe supporting the current way of managing media is really important. Wagtail should support the current way as well as a new way of referencing images/documents/... from pages.

> Would this be an option in the chooser UI? Or perhaps a flag on the image/document itself?

I'm not really sure. But it has to be scoped to pages. I would suggest using some intermediate class type like `Orderable` s.

> The best route is almost always to create a third-party app first, where things can evolve organically, not tied to Wagtail's release process. If you have avenues you'd like to explore, may I suggest trying this approach?

It would be a great honour if I could create such a third-party app for wagtail. But I don't know when that would be possible for me, so don't count on me.

**Comment by ababic** [comment]
Thank you for the thorough reply @HsnVahedi.

Here's a solution that might work... Like we have a 'lock' mechanism for pages to prevent editing, we could technically do the same for images and documents, and lock/unlock permission could be given only to moderators. 

With this in place, you could probably utilise existing hooks to identify and lock the relevant images at the time a page is published (depending on the owner). OR you could maybe specify at a collection-level whether media uploaded to that collection should be locked automatically after uploading. 

What do you think?

**Comment by ababic** [comment]
NOTE: I have removed the bug classification on this issue, as the current design works as intended.

**Comment by HsnVahedi** [comment]
I think that's amazing @ababic 

Letting moderators to lock some of the images/documents referenced by a page solves the problem.

But what about the images/documents that are in the page but are not referenced directly? for example:

```
class SomePage(Page):
    image = models.ForeignKey(
        'wagtailimages.Image', on_delete=models.CASCADE, related_name='+'
    )
    body = StreamField([
        ('image', ImageChooserBlock()),
    ], null=True, blank=True,)
```
To achieve a complete solution, moderators should be able to lock the image blocks in body too.

But I do believe even an incomplete solution (which only lets moderators to lock images/documents that are referenced directly) is enough for many cases.

**Comment by ababic** [comment]
@HsnVahedi Okay, good to know we're on the right track. For clarity, I feel like it may be best to close this issue now and create a separate 'Lockable images and documents' feature request one, linking back here.

**Comment by ababic** [comment]
@HsnVahedi I have also added a point for discussion in the next core team meeting about whether [`CollectionOwnershipPermissionPolicy`](https://github.com/wagtail/wagtail/blob/master/wagtail/core/permission_policies/collections.py#L199) is the right choice image/documents, as the combination of ownership and collection level permissions do conflict with eachother slightly.

**Comment by allcaps** [comment]
Just an idea,

If media is defined on the page with a FileField or ImageField, than that media is tightly bound to the page and falls under Page permissions, Page workflow and Page revisions.

There are some downsides, this media can not be managed in a central place, and ImageField does not support renditions.

Rendition support for ImageField would be useful in many other situations. 

Managing media in central place (or not) and keeping a reference to some source file (or not) are requirements that can differ from project to project. Technically, these scenarios are possible.

**Comment by HsnVahedi** [comment]
Thanks @ababic ! I'm ok to close this issue.

But just one more thing: Why using collections? Just let moderators lock images/documents (or even snippets) exactly the same as locking pages.

**Comment by ababic** [comment]
@HsnVahedi

> But just one more thing: Why using collections? Just let moderators lock images/documents (or even snippets) exactly the same as locking pages.

I think there may be some confusion over what I was suggesting. What I suggested was:
- Locking/unlocking would be done image/document (as you describe)
- Also being able to configure a collection, so that  media uploaded to it was locked automatically on creation

That second suggestion would be complementary, and not necessarily required. Feel free to leave it out of you new ticket if you think it will simplify things. There's no reason it couldn't be done later.

**Comment by ababic** [comment]
@allcaps thanks for adding to the conversation. I rather like this idea:

> Rendition support for ImageField

Would you mind creating a separate ticket for this?

**Comment by HsnVahedi** [comment]
Great! @ababic  I'm closing this issue and waiting for this amazing new feature.

**Comment by HsnVahedi** [comment]
Hello again @ababic ! Wish you the best! Have you got any plans on adding this feature (locking images/documents/...) to wagtail?

**Comment by ababic** [comment]
Hi @HsnVahedi. I thought you were going to create an issue for that new feature request (I haven't seen anything yet). I am personally very busy with client work, so wasn't planning to work on this any time soon. If you put forward a fairly detailed feature request, I think it might make a nice sized feature for a contributor to pick up.

**Comment by HsnVahedi** [comment]
Ok @ababic . I'll do it.
